### PR TITLE
Use GitHub Actions for more OS-specific build tests

### PR DESCRIPTION
**Pull request by kinkie**
Test staged commits using some of our docker images from Jenkins tests.
The added tests (see staged.yaml) were easier to adopt; we are also
working on enabling GitHub Actions for FreeBSD and some other images.

We do not run these added tests for PR commits because existing Ubuntu
tests (defined in pr.yaml) already expose the vast majority of build
problems, and we are worried that running a lot more tests for each PR
push event would consume too much GitHub resources and significantly
increase noise in PR checks summary, obscuring often-easier-to-handle
failures detected by Ubuntu tests.

Also postpone MacOS tests until PR staging. On GitHub, MacOS runners are
x10 more expensive than Linux runners. We use cheaper runners for fast
feedback while still checking MacOS build before each merged commit.

This change does not increase GitHub CI wait time for PR push tests
because those checks are dominated by the unchanged 35min CodeQL-tests
job. However, it reduces total CPU time used (from 2h 30m to 2h) because
we no longer perform 3 MacOS tests for PR push events.

This change doubles GitHub CI wait time for staged commit tests (from
35m to 1h) and drastically increases total CPU time used (from 2h to
17h), primarily due to 84 added docker-based linux-distros checks.

GitHub does not provide any riscv64 runners and free Linux/arm64
runners. Our initial attempts to virtualize Linux/arm64 tests on
MacOS/arm64 runners and to make virtualized FreeBSD and OpenBSD tests
work on Linux/x64 were not successful. Thus, we still rely on Jenkins
for Linux/riscv64, Linux/arm64, FreeBSD/x64, and OpenBSD/x64 tests.

### CODE REVIEW

**Comment by kinkie** [comment]
Note that this is a draft because:
* it still references the branch name to run the tests themselves  - needs to be removed pre-merge
* landing needs to coordinate with changes to anubis' configuration

File: .github/workflows/merge.yaml
Comment (rousskov) [inline .github/workflows/merge.yaml:44]: "latest" feels like the wrong tag. We should use images that were cleared/tested for Squid Project use. The latest image may not be one of those! In other words, how will you test image changes/upgrades without affecting existing/pending PRs if GitHub Actions are supposed to use the _latest_ image?

I am not a docker expert, but please consider something like this instead:

```suggestion
      image: squidcache/buildfarm-${{ matrix.os }}:production
```

or

```suggestion
      image: squidcache/buildfarm-${{ matrix.os }}:stable
```

or even something like this if we foresee using different image variations for master and v6:

```suggestion
      image: squidcache/buildfarm-${{ matrix.os }}:for-master
```

N.B. I do not know what Jenkins is using, but I do not think that should matter for this PR. We do not have to replicate all Jenkins setup problems here.

**Review by rousskov — CHANGES_REQUESTED** [review]
I support migrating Jenkins tests to GitHub. Thank you for working on that!

I hope this partial review allows you to move this PR forward.

> PR description: Github only provides x64 runners and arm64 macos runners. 

GitHub does provide [arm64 Linux runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners/about-larger-runners#specifications-for-general-larger-runners), but they are not free. I tried to correct and expand the above assertion. Please edit further as needed!

I also changed PR title to better reflect current PR limitations.

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:44]: Our current image promotion process calls for :stable. However, opensuse-tumbleweed stable is currently broken, latest is working. I am currently testing latest for promotion. Adding a note to switch to stable before merge

Comment (rousskov) [reply .github/workflows/merge.yaml:44]: AFAICT, we are talking about two unrelated aspects:

* I am talking about the freshest docker image produced by you (i.e. the very last image that was produced). This configuration should not use the freshest image. We should use the image that was _validated_ to work best. In some cases, it will not be the freshest. I assume that "latest" spelling implies "freshest", but perhaps that is not true.
 
* You might be talking about some designation related to OS versions. For example, right now, the latest Ubuntu release is X.Y. This change request concern is _not_ about OS versions and such!
 
Who controls those image suffixes? I assumed it was us.

How will you upgrade from one docker image to another (while staying on the same OS version and just improving/changing something inside the published/shared image)?

**Comment by kinkie** [comment]
> We do not want to virtualize Linux/arm64 tests on MacOS/arm64
> runners

I tried that, it doesn't work.

**Comment by kinkie** [comment]
> Our initial attempts to virtualize FreeBSD and OpenBSD tests on
> Linux/x64 were not successful

FreeBSD and OpenBSD virtualization works just fine.
The test build fails and I haven't had the time to investigate why yet

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:44]: > AFAICT, we are talking about two unrelated aspects:
> 
> * I am talking about the freshest docker image produced by you (i.e. the very last image that was produced). This configuration should not use the freshest image. We should use the image that was _validated_ to work best. In some cases, it will not be the freshest. I assume that "latest" spelling implies "freshest", but perhaps that is not true.

That is, indeed, correct. "latest" is the latest image that I built and pushed to the docker hub.
I re-tag it as "stable" after I successfully test no regressions against the current test set

> * You might be talking about some designation related to OS versions. For example, right now, the latest Ubuntu release is X.Y. This change request concern is _not_ about OS versions and such!

Yes, I had understood that

> Who controls those image suffixes? I assumed it was us.

It is us.

> How will you upgrade from one docker image to another (while staying on the same OS version and just improving/changing something inside the published/shared image)?

that suffix is just a label. "latest" is by convention the most recently pushed version.
Our own promotion process is driven by [this makefile rule](https://github.com/kinkie/dockerfiles/blob/11eb9feb5cc0fc92fc091d84932a8a0af06f0abe/Makefile#L106):

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:183]: Done

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:114]: The intent behind leaving the code in is as documentation and as an example.
If we ever will want to disable some combination, it's not entirely trivial how to do it, and I'd rather have a comment than spend hours reverse engineering how the exclusion works

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml]: removed

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:46]: Interesting. It was working when I tried in my own repository

**Comment by rousskov** [comment]
> > We do not want to virtualize Linux/arm64 tests on MacOS/arm64
> > runners
 
> I tried that, it doesn't work.

OK, I adjusted the claim to reflect that failed attempt. If you are saying that it cannot work (for some fundamental reason), then further adjustments are needed. FWIW, I was told it can work, in principle, but this is not my area of expertise, and I have not studied the issue myself.

> > Our initial attempts to virtualize FreeBSD and OpenBSD tests on
> > Linux/x64 were not successful
> 
> FreeBSD and OpenBSD virtualization works just fine. The test build fails and I haven't had the time to investigate why yet

OK, I think that outcome was covered by the original "virtualize _tests_" phrase, but I adjusted that text to clarify.

Both phrases or their better versions are needed to explain why we are not migrating the rest of the Jenkins tests: We cannot say "A is not available. _Thus_, we are not testing on A, B, C, and D."

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:46]: removed, and saved in PR #1919 for followups

**Comment by kinkie** [comment]
> > > We do not want to virtualize Linux/arm64 tests on MacOS/arm64
> > > runners
> 
> > I tried that, it doesn't work.
> 
> OK, I adjusted the claim to reflect that failed attempt. If you are saying that it cannot work (for some fundamental reason), then further adjustments are needed. FWIW, I was told it can work, in principle, but this is not my area of expertise, and I have not studied the issue myself.

The main reason is that the MacOS runners do not come with docker installed.
There is documentation on how to do it online, but the one I tried didn't work.
I'd rather focus on getting the matrix out, then iterate on this part if it works.

> > > Our initial attempts to virtualize FreeBSD and OpenBSD tests on
> > > Linux/x64 were not successful
> > 
> > 
> > FreeBSD and OpenBSD virtualization works just fine. The test build fails and I haven't had the time to investigate why yet
> 
> OK, I think that outcome was covered by the original "virtualize _tests_" phrase, but I adjusted that text to clarify.
> 
> Both phrases or their better versions are needed to explain why we are not migrating the rest of the Jenkins tests: We cannot say "A is not available. _Thus_, we are not testing on A, B, C, and D."

Sounds good to me

File: .github/workflows/merge.yaml
Comment (rousskov) [reply .github/workflows/merge.yaml]: > "latest" is the latest image that I built and pushed to the docker hub.
> I re-tag it as "stable" after I successfully test no regressions against the current test set

OK, then master (i.e. this PR) should use "stable" tag. You can use "latest" tag in your Squid repository to check new images on GitHub (if possible, I would avoid using Jenkins for those checks).

**Comment by rousskov** [comment]
>  I'd rather focus on getting the matrix out, then iterate on this part if it works.

I fully support starting with what works today. I just want us to be clear how we decided what to include -- that information cannot be guessed by looking at future commit diff...

File: .github/workflows/pr.yaml
Comment (rousskov) [reply .github/workflows/pr.yaml:114]: > The intent behind leaving the code in is as documentation and as an example.
> I'd rather have a comment than spend hours reverse engineering how the exclusion works

This example is already in master/v7. It will be there forever, even after you delete it here. Yes, it might take a minute to find, but that is not really reverse engineering. Moreover, commented out code tends to get stale, requires review/maintenance, and is in the way of other refactoring.

And where do we draw the line? You leave GitHub Actions configuration examples. I add used-to-work C++ code examples. Alice starts fighting for leaving buggy C++ code examples, so that he does not repeat the same mistake twice. Bob starts copying those.... Let's simply remove unused stuff instead.

Comment (rousskov) [reply .github/workflows/pr.yaml:114]: > Alex: Why? How are we going to decide what tests to run for PR branch and what tests to run for staged commits? Please document this important answer in PR description, not (just) this change request comment. We need to reach consensus on test placement scheme before we know which of the two workflow files each test belongs to...

Thank you for updating PR description to answer this question. I edited that answer a little. I will come back to that answer later, to check whether we should move more PR commit tests to PR staging phase, for the same "GitHub cost saving" reason.

**Review by rousskov — CHANGES_REQUESTED** [review]
This is another incremental review to help you make progress.

> Looks ready for review and merge

GitHub seems to violently disagree:

```
The workflow is not valid. .github/workflows/pr.yaml (Line: 106, Col: 17): Unexpected value ''
```

Please fully test before requesting the next review.

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:1]: Why not "auto.yaml" and "pr.yaml"? 

"default" makes no sense as soon as another file appears, and it is questionable it ever did
"staged" might make sense
"*-branch" is redundant

In the end, the names really do not matter to github, and carry very little relevance to humans as well, it's not something we will touch on any cadence as soon as we have a good test coverage

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:1]: I disagree on characterizing "default" as desciptive.

I propose "auto" and "pr" which details the triggers for the workflow as in [another thread](https://github.com/squid-cache/squid/pull/1917/files#r1803832450)

Also, there is not much benefit in keeping the different jobs currently in pr.yaml in a single file , the only reused part is the definition of two environment variables, of which one is not even used everywhere

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:60]: ok. I'll do it the other way around tho.

--verbose means that the output of the build is visible in the worfklow execution window, no need to download artifacts
--use-config-cache means that recursive invocations of ./configure (e.g. in libltdl) do not need to rerun shared tests. config.cache does not survive workflow executions anyway

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:72]: Why? This all is unnecessary complication, it buys us nothing at all.
We do not automate artifacts collection. If we ever will want a machine
to do touch these artifacts, that will be the moment to consider
whether to standardise.

Also, this workflow is working just fine, I wouldn't characterise any of the
suggestions in this comment as "problems".

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml:69]: Why? It is a single test for now, so -test is appropriate.
If we will ever evolve to a combinations matrix, then sure

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:1]: sure

**Comment by kinkie** [comment]
The workflow is now tested, see https://github.com/squid-cache/squid/actions/runs/11393594273

File: .github/workflows/merge.yaml
Comment (rousskov) [reply .github/workflows/merge.yaml:1]: > Francesco: Why not "auto.yaml" and "pr.yaml"?

Not sure where "Why not" is coming from -- I did not say anything about this combination of names in my change request. I am interpreting this as a "How about going with suggested option 2 but dropping `-branch` suffixes?" question.

>  "*-branch" is redundant

I disagree because, without `-branch`, "auto" will mean something like "automatically applied" or "applied by default" to many readers while "auto-branch" will significantly reduce that problematic reader set -- folks will be more likely to discover the existence of a _branch_ named "auto". It is a fairly standard/common branch name -- other projects use "auto" for staged tests -- but not standard/common enough to expect correct interpretation in this context IMO.

"pr.yaml" has the same problems as discussed in option 2's "pr-branch.yaml": It is somewhat misleading because pr.yaml is used for testing (merged) master commits as well (and not just PR branch commits). Since we cannot find problem-free names, we can decide that these problems are not worth solving.

I still recommend adding "-branch" but I do not insist on it. If you insist on not using "-branch", then I suggest "staged.yaml" and "pr.yaml".

> "default" makes no sense as soon as another file appears

I disagree. I have already detailed why "default" works more-or-less OK IMO, but I do not insist on us preserving default.yaml name. If you do not like it, let's just cross it out from the list of candidates.


> [The names is] not something we will touch on any cadence as soon as we have a good test coverage

I expect us to continue adding workflows, but we do not need to agree on that speculation.

File: .github/workflows/merge.yaml
Comment (rousskov) [reply .github/workflows/merge.yaml]: > I disagree on characterizing "default" as desciptive.

I did not use chat characteristic: The "names" this change request talks about are the values of the YAML `name:` attribute, not filenames. I had to use (some) file names to attribute suggested values to the right file.

`name:` values in the current PR still need to be fixed: We probably do not want to waste valuable space on "GitHub CI" prefixes -- they are obvious in contexts where these names are used by GitHub. We should mention that pr.yaml is used for more than PR commit checks. I suggest:

* staged.yaml: Staged commit checks
* pr.yaml: PR commit checks and PR target branch commit checks

**Review by rousskov — CHANGES_REQUESTED** [review]

File: .github/workflows/auto.yaml
Comment (rousskov) [inline .github/workflows/auto.yaml:52]: If these two lines were added by accident, please remove them. AFAICT, they do not exist in the relevant parts of the official sources (i.e. in default.yaml instructions for `build-tests` jobs[^1]), and they force [actions/checkout](https://github.com/actions/checkout?tab=readme-ov-file#checkout-v4) to do more work than necessary for our build tests: "Only a single commit is fetched _by default_".

The same concern applies to MacOS tests, of course -- this part of the code is duplicated in the current PR.

If all extra depth configurations is removed from this file, please remove `CHECKOUT_FETCH_DEPTH` declaration as well (as unused).

[^1]: Fetching extra Squid repository commits is necessary for our `source-maintenance-tests`, but this new workflow does not run those tests.

Comment (kinkie) [reply .github/workflows/auto.yaml]: Got you. I was wandering in fact; the code that uses part of squid's commits is not in the github actions but in the scripts.
It wasn't clear to me in the comments in the current master code.
Removed.

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml]: "GitHub CI" is what is currently used, since commit 2ed8cc427ff .
It was true then, it's worth now. Adopting in order to progress

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml]: I struggle to see what could be controversial about them, but we can have that discussion in a followup PR

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml]: This feels like picking nits.
In order to clarify what is happening and avoid redundancy, I have renamed the tests and used "test-builds" as id.

Please let's progress unless it is really important or misleading.

File: .github/workflows/merge.yaml
Comment (kinkie) [reply .github/workflows/merge.yaml]: Ah right.
The answer is pretty simple then: simplicity.
The setup phase on homebrew is quite different from Linux's.
As this PR adds significant complexity by having a large number of matrix combinations, I decided it was best to isolate MacOS complexity on its own.
We could possibly parallelise macos tests but I see little benefit in doing that for now.

In another thread you mentioned that you assume there will be more platforms tested, and that is indeed correct: the FreeBSD workflow is working and will can discuss it as soon as this PR lands

File: .github/workflows/staged.yaml
Comment (rousskov) [inline .github/workflows/staged.yaml:114]: AFAICT, this log name is used in GitHub Actions artifacts index that also contains log name from docker-based jobs and, hence, it should identify this job or OS (at least):

```suggestion
          name: build-logs-macos
```

**Review by rousskov — CHANGES_REQUESTED** [review]
I edited PR description to remove stale "full-blown" claim and explain why we delay docker-based tests. Please adjust further as needed.

Please mention how many new tests we are adding (for a PR push event and, separately, for a staged commit) and what is the approximate increase in total test time (for a PR push event and, separately, for a staged commit). These stats would help document the impact of this PR changes.

As for YAML changes, I found one [serious problem](https://github.com/squid-cache/squid/pull/1917#discussion_r1812927141) and a couple of lesser naming problems. Please see new change requests for details.

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:1]: I don't think we want to remove the name field. I agree to make it shorter.

We can use comments in the yaml file to clarify intent; they will not appear in the UI but they will be visible in the source tree

I don't think we need to identify with "GH"; after all the directory these live in is .github/workflows; it is quite obvious that any file in there is a GitHub workflow.

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:12]: I think I understand what you mean. My understanding was that
since we test the PR when it is updated, there is no need to re-run
these tests since they've been already run.
But then this would miss running these tests after any other PR
has been rebased before this. Good point, re-adding the "auto"
branch here is the easiest, safest, and least expensive option

File: .github/workflows/staged.yaml
Comment (kinkie) [reply .github/workflows/staged.yaml]: Makes sense. Adding.

File: .github/workflows/pr.yaml
Comment (rousskov) [reply .github/workflows/pr.yaml]: > My understanding was that since we test the PR when it is updated, there is no need to re-run these tests ... But then this would miss running these tests after any other PR has been rebased before this.

If you replace "rebased" with "merged", then this would be more-or-less accurate. Here is a simplified summary:

* The latest PR branch commit X is a combination of master branch _as it existed at PR branching time_ and all PR branch commits. Our current CI does not test these commits, but we often talk as if it does.
* So called GitHub "merge commit" Y is a combination of master branch _as it existed at the last PR branch update time_ IIRC and all PR branch commits. Jenkins and GitHub Actions test these commits when they test "PR".
* A staged commit Z is a combination of master branch _as it existed when Z was staged_[^1] and all PR branch commits. Jenkins and GitHub Actions test these commits when they test "auto" branch.

[^1]: Z's parent commit is then-current master. If the corresponding PR is merged, master becomes Z.

Thus, the relevant difference here is the difference between Y and Z. That difference is all commits that happened on master since PR branch was last updated. If a given PR stays constant, that difference grows with every merged PR.

File: .github/workflows/pr.yaml
Comment (rousskov) [reply .github/workflows/pr.yaml]: > I don't think we need to identify with "GH"; after all the directory these live in is .github/workflows; it is quite obvious that any file in there is a GitHub workflow.

We want GH (or equivalent) because these names are used as entries in an index that includes both GitHub and non-GitHub checks:

* PR commit / build-tests(ubuntu-22.04,clang,minimal) (pull_request)  Successful in 7m
* PR commit / build-tests(ubuntu-22.04,clang,maximus) (pull_request)  Successful in 15m
* Jenkins(build test)  Pending

In the above sample PR page output, GitHub and Jenkins both test "PR commit"[^1]. it would be good to add "GH" (or equivalent) just like Jenkins adds "Jenkins". Doing so will improve naming consistency and may reduce the number of folks that do not know where non-Jenkins checks are coming from.

[^1]: Well, they actually do not test "PR commit" -- they test [commit Y not X](https://github.com/squid-cache/squid/pull/1917#discussion_r1813340507)) -- but this part of the discussion is about "GH" prefix, not a misleading (especially for staged tests!) "PR commit" description that follows that prefix.

File: .github/workflows/staged.yaml
Comment (rousskov) [inline .github/workflows/staged.yaml:10]: The suite of tests in this file is not "full" or "comprehensive" AFAICT -- it excludes functionality tests, source maintenance tests, CodeQL tests, and possibly some of Jenkins tests. 

The concern about Anubis configuration update applies to the tests defined in the other YAML file as well AFAICT because those tests are also executed for staged commits.

Similar to the other change request, I recommend removing this misleading description. We can make progress by documenting our changing test placement logic in PR description, at least for now. Once the dust settles, all files are (re)named appropriately, all tests are placed appropriately, we can decide to document our finalized logic in sources.

```suggestion
```

FWIW, right now, I suspect that this file might become something like "extra.yaml" or even "slow.yaml" while the old one might become something like "minimal.yaml" or "quick.yaml" (or even retain its original default.yaml name!).

**Review by rousskov — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:7]: My understanding is that for each workflow file, all jobs defined in the jobs: section get executed for all triggers defined in the on: section, barring any job-specific conditions.

Polishing is always possible, but to me some clarification of the intent is better than trying to express it three words of job name.

Comment (kinkie) [reply .github/workflows/pr.yaml:7]: Tried rewording and polishing. Moved the comment to the top of the file, clarifying
that it references to all jobs defined in the file

File: .github/workflows/staged.yaml
Comment (kinkie) [reply .github/workflows/staged.yaml]: > The suite of tests in this file is not "full" or "comprehensive" AFAICT -- it excludes functionality tests, source maintenance tests, CodeQL tests, and possibly some of Jenkins tests.

Lies, big lies, and carefully chosen metrics on my side.
I'm better clarifying intent that the aim is to check for compiler, OS, and platform
portability issues.

> The concern about Anubis configuration update applies to the tests defined in the other YAML file as well AFAICT because those tests are also executed for staged commits.

Yes. Clarified as well, copying that snippet to all comments.

> Similar to the [other change request](https://github.com/squid-cache/squid/pull/1917#discussion_r1813371727), I recommend removing this misleading description. We can make progress by documenting our changing test placement logic in PR description, at least for now. Once the dust settles, all files are (re)named appropriately, all tests are placed appropriately, we can decide to document our finalized logic in sources.
> 
> FWIW, right now, I suspect that this file might become something like "extra.yaml", "more.yaml", or even "slow.yaml" while the old one might become something like "first.yaml" or "quick.yaml" (or even retain its original default.yaml name!).

Possibly. Whatever it will be, I don't think it's worth delaying the PR on.
As soon as this goes in, I can start merging FreeBSD support which is ready and waiting, and then look into ccache which is mostly ready and only needs
to be fixed in the context of the squid-cache account, and then more platforms which
I have some idea on how to do but requires some tinkering.
When that is done, we can - for the most part if not in full - retire jenkins.

File: .github/workflows/pr.yaml
Comment (rousskov) [inline .github/workflows/pr.yaml:1]: Please rename this to first.yaml or quick.yaml instead. This workflow defines tests that are executed for non-PR commits (as well as PR commits) so calling it pr.yaml is quite misleading (and this PR had related bugs that we had to fix). The primary purpose of this workflow is to provide quick/initial feedback to PR authors. Let's reflect that purpose via a (rather vague but less misleading) filename.

**Review by rousskov — DISMISSED** [review]
I hope we are one or two iterations away from merging this PR. The discussion regarding workflow split/purpose appears to have converged on common understanding. I have also resolved all old change requests except for two; please do not forget about those two (they can be accessed via `Conversations` dialog at the top of `Files changed` tab).

File: .github/workflows/merge.yaml
Comment (rousskov) [reply .github/workflows/merge.yaml]: > I would even use os/compiler/layer matrix (with two components fixed) so that we can use the same artifact name scheme and naming code instead of unnecessary special-casing MacOS in that way.

Just to avoid misunderstanding: This change request does _not_ ask to merge MacOS and non-MacOS tests into _one_ common matrix (and `if` statements to differentiate OSes). MacOS tests should use their own matrix.

To recap: If docker tests iterate over build layers (among other matrix dimensions), then MacOS tests should also iterate over build layers (at least). Otherwise (i.e. if docker tests are changed to stop iterating over build layers), MacOS tests should not iterate over build layers either.

File: .github/workflows/pr.yaml
Comment (kinkie) [reply .github/workflows/pr.yaml:1]: Can do

File: .github/workflows/staged.yaml
Comment (kinkie) [reply .github/workflows/staged.yaml:1]: "extra" is non descriptive; "more" even less so. "portability" could capture the intent of the file, and it works
well to also contain upcoming FreeBSD tests

**Comment by kinkie** [comment]
Current status: simulated staged commit tests has 103 [checks](https://github.com/kinkie/squid/pull/12/checks) in total :
- 85 from "staged commit" - completed in 1h7m and using 16h CPU time
- 9 from "PR commit" - completed in 1h33m and using 2h CPU time, run twice (one due to PR trigger and one due to push)

File: .github/workflows/staged.yaml
Comment (rousskov) [reply .github/workflows/staged.yaml:1]: > "extra" is non descriptive; "more" even less so.

Both words describe exactly what we want IMO: Tests in addition to tests present in quick.yaml.

> "portability" could capture the intent of the file

Portability is not a good choice because portability tests are found in quick.yaml as well.

Comment (kinkie) [reply .github/workflows/staged.yaml:1]: I disagree. These tests are not in addition to quick.yaml, they test a different thing.

I see three ways forward:
- leave as is. We are discussing nits
- one file - one job. This will make intent easier to express in a word. Expanding scope, for the PR, unfortunately. Might be a followup one to avoid that, but then we should land this as is
- have "fast.yaml" and "expensive.yaml" or "slow.yaml"

**Review by rousskov — COMMENTED** [review]
> Current status: simulated staged commit tests has 103 [checks](https://github.com/kinkie/squid/pull/12/checks) in total ...

Thank you. I have added a couple of paragraphs to PR description summarizing those stats. Please check and adjust as needed.

Other than that PR description check/adjustment, only portability.yaml renaming issue is left AFAICT. I am not going to block this PR based on that disagreement alone, but I do encourage you to rename.

File: .github/workflows/staged.yaml
Comment (rousskov) [reply .github/workflows/staged.yaml:1]: > I disagree. These tests are not in addition to quick.yaml, they test a different thing.

Tests in file A are in addition to (different) tests in file B. It is a basic fact. We should not be arguing about it.

> * leave as is. We are discussing nits

Misleading name is not a nit, but I do not insist on improving this name.
 
> * one file - one job. This will make intent easier to express in a word.

I would not go there (unless needed for other reasons) because more files make the results even more difficult to find on GitHub Actions page. Even two files make navigation there quite tricky. There is also some configuration code duplication in one-job-per-file approach.

> * have "fast.yaml" and "expensive.yaml" or "slow.yaml"

I think slow.yaml is much better than portability.yaml and slightly better than expensive.yaml. If you are OK with slow.yaml, let's settle on that (for now).

**Comment by kinkie** [comment]
Ok, should be good to go as soon as you're happy to press the button, @rousskov

File: .github/workflows/quick.yaml
Comment (rousskov) [inline .github/workflows/quick.yaml:8]: This workflow also tests master and staged commits. None of them are really "PR commit"s. There was an old change request about that somewhere.

```suggestion

name: GH quick tests
```

or

```suggestion

name: Quick tests
```

or

```suggestion
```


The first and last option would need the corresponding slow.yaml update, but there is a dedicated change request for that file.

The last one is my favorite.

**Comment by kinkie** [comment]
After this lands, I'll take care of disconnecting jenkins x64 checks and realigning Anubis configuration, then backport to v6 and doing the same there; then look into freebsd and ccache support

**Comment by rousskov** [comment]
> Ok, should be good to go as soon as you're happy to press the button, @rousskov

I am not happy, but I am not blocking this PR.

**Pull request closed without merging**
